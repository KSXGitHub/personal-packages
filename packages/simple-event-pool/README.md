# Simple Event Pool

Set up a simple event pool

## Usage

### Create

```javascript
const { createEventPool, some, none } = require('@khai96x/simple-event-pool')

const target = createEventPool({
  setInterval,
  clearInterval,
  delay: 1000 // will call triggers once for every 1000ms
})
  .createAutoTrigger('never', () => none())
  .createAutoTrigger('every second', param => some(param.iterationCount))
  .createAutoTrigger(
    'every 3 seconds',
    param => param.iterationCount % 3 === 0
      ? some(param.iterationCount)
      : none()
  )
  .createAutoTrigger(
    'every 15 seconds',
    param => param.iterationCount % 15 === 0
      ? some(param.iterationCount)
      : none()
  )
  .startEventLoop() // start the interval loop
```

### Add Listener

```javascript
target
  .addListener('never', () => console.log('this message will never show'))
  .addListener('every second', value => console.log('1x', value))
  .addListener('every 3 seconds', value => console.log('3x', value))
  .addListener('every 15 seconds', value => console.log('15x', value))
  .addListener('every 15 seconds', () => target.stopEventLoop())
```

**Expected Output:**

```
1x 1
1x 2
1x 3
3x 3
1x 4
1x 5
1x 6
3x 6
1x 7
1x 8
1x 9
3x 9
1x 10
1x 11
1x 12
3x 12
1x 13
1x 14
1x 15
3x 15
15x 15
```
