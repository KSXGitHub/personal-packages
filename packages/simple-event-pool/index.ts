import { Option, none, some } from '@tsfun/option'
import AdvMapInit from 'advanced-map-initialized'

type MaybePromise<Value> = Value | Promise<Value>
type OptionPromise<Value> = MaybePromise<Option<Value>>

export interface SetIntervalFunc<ID> {
  (func: () => void, delay: number): ID
}

export interface ClearIntervalFunc<ID> {
  (id: ID): void
}

export interface EventPoolOptions<IntervalID> {
  readonly setInterval: SetIntervalFunc<IntervalID>
  readonly clearInterval: ClearIntervalFunc<IntervalID>
  readonly delay: number
}

export interface EventLoop {
  set (): this
  clear (): this
}

export interface EventTarget<Info, ID> {
  addListener (event: ID, listener: (info: Info) => void): this
  removeListener (event: ID, listener: (info: Info) => void): this
}

export interface EventAutoTrigger<Info, ID> extends EventTarget<Info, ID> {
  createAutoTrigger<Info, ID> (event: ID, check: EventChecker<Info>): EventAutoTrigger<Info, ID> & this
}

export interface EventChecker<Info> {
  (count: number): OptionPromise<Info>
}

export type EventPool<Info, ID> = EventLoop & EventAutoTrigger<Info, ID>

export function createEventPool<IntervalID> (options: EventPoolOptions<IntervalID>) {
  const { setInterval, clearInterval, delay } = options

  let id: Option<IntervalID> = none()

  function set () {
    if (id.tag) throw new Error('Interval already set')
    id = some(setInterval(intervalCallback, delay))
    return pool
  }

  function clear () {
    if (!id.tag) return pool
    clearInterval(id.value)
    id = none()
    return pool
  }

  // id -> checker
  const checkers = new Map<any, EventChecker<any>>()

  function createTrigger<Info, ID> (id: ID, check: EventChecker<Info>) {
    checkers.set(id, check)
    return pool
  }

  // id -> [listener]
  const listeners = new AdvMapInit<any, Set<(info: any) => void>>(Map, () => new Set())

  function addListener (event: any, listener: any) {
    listeners.get(event).add(listener)
    return pool
  }

  function removeListener (event: any, listener: any) {
    listeners.get(event).delete(listener)
    return pool
  }

  let count = 0

  function intervalCallback () {
    for (const [id, check] of checkers) {
      void Promise.resolve(check(count))
        .then(opt => opt.tag && callListeners(id, opt.value))

      count += 1
    }
  }

  function callListeners (id: any, info: any) {
    for (const fn of listeners.get(id)) {
      fn(info)
    }
  }

  const pool: EventPool<never, never> = {
    set,
    clear,
    createAutoTrigger: createTrigger,
    addListener,
    removeListener
  }

  return pool
}

export { Option, some, none }
