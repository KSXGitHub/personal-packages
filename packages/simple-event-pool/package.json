{
  "name": "@khai96x/simple-event-pool",
  "version": "0.0.1",
  "description": "Set up a simple event pool",
  "author": "Hoàng Văn Khải <hvksmr1996@gmail.com>",
  "license": "MIT",
  "homepage": "https://github.com/KSXGitHub/personal-packages#readme",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/KSXGitHub/personal-packages.git"
  },
  "bugs": {
    "url": "https://github.com/KSXGitHub/personal-packages/issues"
  },
  "dependencies": {
    "@tsfun/option": "^0.0.1",
    "advanced-map-initialized": "0.0.2",
    "tslib": "^1.10.0",
    "@types/node": "^12.6.8"
  }
}
