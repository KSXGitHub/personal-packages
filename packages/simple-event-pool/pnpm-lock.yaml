dependencies:
  '@tsfun/option': 0.0.1
  '@types/node': 12.6.8
  advanced-map-initialized: 0.0.2
  tslib: 1.10.0
lockfileVersion: 5.1
packages:
  /@tsfun/option/0.0.1:
    dependencies:
      '@tsfun/prv-option-result-common': 0.0.1
      '@types/node': 11.13.18
      tslib: 1.10.0
    dev: false
    engines:
      node: '>= 8.9.0'
    resolution:
      integrity: sha512-o6vnNuR5gTM+fujbB948flvKN+a/6NIWS/toEzyzd114G8jLptM9uTsDJlf/ZtkbXpRawp9CJm4pRf8sEMib1w==
  /@tsfun/prv-option-result-common/0.0.1:
    dependencies:
      '@types/node': 11.13.18
      tslib: 1.10.0
    dev: false
    engines:
      node: '>= 8.9.0'
    resolution:
      integrity: sha512-FI4Zcc3I4lal4cMdBpWp0O86Ep4G/o8Q7cD47qLXaLa5kom+6pYRO4+XYBjSLKK+2h6rGWd1IvILmmMyAGRN3g==
  /@types/node/10.14.14:
    dev: false
    resolution:
      integrity: sha512-xXD08vZsvpv4xptQXj1+ky22f7ZoKu5ZNI/4l+/BXG3X+XaeZsmaFbbTKuhSE3NjjvRuZFxFf9sQBMXIcZNFMQ==
  /@types/node/11.13.18:
    dev: false
    resolution:
      integrity: sha512-sMHdXLccscaxI+Hlzz58yLQue3lQyXP+6aQniDRi5oju3n0123kcjhicVGF20WF7cHiwJ2vxMbXc4skOclosoA==
  /@types/node/12.6.8:
    dev: false
    resolution:
      integrity: sha512-aX+gFgA5GHcDi89KG5keey2zf0WfZk/HAQotEamsK2kbey+8yGKcson0hbK8E+v0NArlCJQCqMP161YhV6ZXLg==
  /advanced-map-base/0.0.2:
    dependencies:
      '@types/node': 10.14.14
      advanced-map-types: 0.0.2
      tslib: 1.10.0
      typescript-miscellaneous: 0.1.1
    deprecated: Depends on wrong version of "convenient-typescript-utilities"
    dev: false
    engines:
      node: '>= 8.9.0'
    resolution:
      integrity: sha512-fj4I4dNaEnVl9bSKsMXtgyVo4+kgV04ScFyLjyIfdNWp1FnRsCzWHtZDGblgq/Nau/2EMegOqW+PDN9qKmGDIA==
  /advanced-map-initialized/0.0.2:
    dependencies:
      '@types/node': 10.14.14
      advanced-map-base: 0.0.2
      advanced-map-types: 0.0.2
      tslib: 1.10.0
    dev: false
    engines:
      node: '>= 8.9.0'
    resolution:
      integrity: sha512-L6/yZylpOElrSpsZTPiMvpCsgnybX8JxYZX88EO2h+yNHYoqwRCojjNOifKSuoIFiWUb5mPFA0tlW7oTM8iOiQ==
  /advanced-map-types/0.0.2:
    dependencies:
      '@types/node': 10.14.14
      tslib: 1.10.0
      typescript-miscellaneous: 0.1.1
    deprecated: Depends on wrong version of "convenient-typescript-utilities"
    dev: false
    engines:
      node: '>= 8.9.0'
    resolution:
      integrity: sha512-L6IJYPR2uFyt6W/R+1j70pG1M0KzkY9TpV0ySkJuhRAhOFxc2DZ3+0621eaivFoF5Yyc2+dQDwwpRUBwwa49iw==
  /tslib/1.10.0:
    dev: false
    resolution:
      integrity: sha512-qOebF53frne81cf0S9B41ByenJ3/IuH8yJKngAX35CmiZySA0khhkovshKK+jGCaMnVomla7gVlIcc3EvKPbTQ==
  /typescript-miscellaneous/0.1.1:
    dependencies:
      '@types/node': 10.14.14
      tslib: 1.10.0
    dev: false
    engines:
      node: '>= 8.9.0'
      typescript: '>= 3.0.0'
    resolution:
      integrity: sha512-dyGMilZPOL/qQMlJSWK+SbQyJpf6YZZPmPrbbdHGkAvGUvlhv4kGKTgddXWLxFWHTirR2I0VfD8mSwia3WLGtA==
specifiers:
  '@tsfun/option': ^0.0.1
  '@types/node': ^12.6.8
  advanced-map-initialized: ^0.0.2
  tslib: ^1.10.0
