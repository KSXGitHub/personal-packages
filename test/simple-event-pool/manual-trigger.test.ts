import { createEventPool } from '@khai96x/simple-event-pool'

describe('one event', () => {
  const EVENT = Symbol('EVENT')
  type Info = 'a' | 'b' | 'c'

  const pool = createEventPool({
    setInterval,
    clearInterval,
    delay: 0
  })
    .createManualTrigger<Info, typeof EVENT>()

  describe('add listener but do not trigger', () => {
    it('does not call listener', () => {
      const listener = jest.fn()
      pool.addListener(EVENT, listener)
      expect(listener).not.toBeCalled()
    })
  })

  describe('add listener then trigger', () => {
    function setup () {
      const listener = jest.fn()

      pool
        .addListener(EVENT, listener)
        .trigger(EVENT, 'a')
        .trigger(EVENT, 'b')
        .trigger(EVENT, 'c')

      return { listener }
    }

    it('calls listener for every trigger', () => {
      const { listener } = setup()
      expect(listener).toBeCalledTimes(3)
    })

    it('calls listener with expected arguments', () => {
      const { listener } = setup()
      expect(listener.mock.calls).toMatchSnapshot()
    })
  })

  describe('add listener, then remove listener, and then trigger', () => {
    function setup () {
      const listener = jest.fn()

      pool
        .addListener(EVENT, listener)
        .trigger(EVENT, 'a')
        .trigger(EVENT, 'b')
        .removeListener(EVENT, listener)
        .trigger(EVENT, 'c')

      return { listener }
    }

    it('calls listener for every trigger', () => {
      const { listener } = setup()
      expect(listener).toBeCalledTimes(2)
    })

    it('calls listener with expected arguments', () => {
      const { listener } = setup()
      expect(listener.mock.calls).toMatchSnapshot()
    })

    it('does not call listener after removal', () => {
      const { listener } = setup()
      expect(listener).not.toBeCalledWith('c')
    })
  })
})
