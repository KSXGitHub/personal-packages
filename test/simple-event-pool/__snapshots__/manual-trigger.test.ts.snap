// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`one event add listener then trigger calls listener with expected arguments 1`] = `
Array [
  Array [
    "a",
  ],
  Array [
    "b",
  ],
  Array [
    "c",
  ],
]
`;

exports[`one event add listener, then remove listener, and then trigger calls listener with expected arguments 1`] = `
Array [
  Array [
    "a",
  ],
  Array [
    "b",
  ],
]
`;
